### PR TITLE
feat: LeanSAT's LRAT parsers + SAT solver interface

### PR DESCRIPTION
**Pull request by hargoniX**
Step 5/6 in upstreaming LeanSAT.

### CODE REVIEW

File: src/Lean/Elab/Tactic/BVDecide/External.lean [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+    if ← tk.isSet then
+      interrupted
+      throw <| .internal Core.interruptExceptionId
+    else
+      x
+
+/--
+Call the SAT solver in `solverPath` with `problemPath` as CNF input and ask it to output an LRAT
+UNSAT proof (binary or non-binary depending on `binaryProofs`) into `proofOutput`. To avoid runaway
+solvers the the solver is runwith `timeout` in seconds as a maximum time limit to solve the problem.
```
Comment (tobiasgrosser) [inline src/Lean/Elab/Tactic/BVDecide/External.lean:122]:
```suggestion
solvers the solver is runwith `timeout` in seconds as a maximum time limit to solve the problem.
```

File: src/Lean/Elab/Tactic/BVDecide/LRAT/Parser.lean [hunk 2]
```diff
@@ -0,0 +1,428 @@
...
+    (pivotInt.natAbs, false)
+
+
+open Lean.Parsec
+open Lean.Parsec.ByteArray
+
+namespace Text
+
+/-
+This implements a (corrected) version of the grammar presented in:
```
Comment (tobiasgrosser) [inline src/Lean/Elab/Tactic/BVDecide/LRAT/Parser.lean:37]: I wonder if it would make sense to not the corrections you applied here.

**Review by tobiasgrosser — COMMENTED** [review]
Nice. 😃

**Review by TwoFX — APPROVED** [review]

File: src/Lean/Elab/Tactic/BVDecide/LRAT/Parser.lean [hunk 3]
```diff
@@ -0,0 +1,431 @@
...
+partial def parseActions : Parser (Array IntAction) := do
+  go #[]
+where
+  go (actions : Array IntAction) : Parser (Array IntAction) := do
+    let action ← parseAction
+    let actions := actions.push action
+    if ← eof? then
+      return actions
+    else
+      go actions
```
[the remark this answers is not in the document]
Comment (hargoniX) [reply src/Lean/Elab/Tactic/BVDecide/LRAT/Parser.lean:270]: Yes, so that we don't accidentally end up missing something due to a parser mistake.

**Pull request merged**